### PR TITLE
Rust diesel libmysqlclient

### PR DESCRIPTION
**Pull request by FinnDore**
## What does this PR address?
Adds tests for the issue detailed in https://github.com/railwayapp/nixpacks/issues/328. Ive added the test and would also like to fix the underlying issue in this pr, but need assistance with this part.

- [X] Make sure to follow [GitHub's guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on creating PR.
- [X] Did you read through [contribution guidelines](https://github.com/railwayapp/nixpacks/blob/main/CONTRIBUTING.md)?
- [ ] Did your changes require updates to the documentation? Have you updated those accordingly?
- [ ] Did you write tests to cover your changes? Did it pass locally when running `cargo test`?
- [ ] Have you run `cargo fmt`, `cargo check` and `cargo clippy` locally to ensure that CI passes?

### CODE REVIEW

**Comment by Milo123459** [comment]
Could make this work for Postgres and other databases diesel supports in this pr?

File: examples/rust-diesel-mysql/.gitignore [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# Edit at https://www.toptal.com/developers/gitignore?templates=rust
+
+### Rust ###
+# Generated by Cargo
+# will have compiled files and executables
+/target/
+
+# Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
+# More information here https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
+Cargo.lock
```
Comment (Milo123459) [inline examples/rust-diesel-mysql/.gitignore:11]: Can you just copy a gitignore from another rust example

File: tests/docker_run_tests.rs [hunk 2]
```diff
@@ -222,6 +223,53 @@ fn run_postgres() -> Container {
     }
 }
 
+fn run_mysql() -> Container {
+    let mut docker_cmd = Command::new("docker");
```
Comment (Milo123459) [inline tests/docker_run_tests.rs:227]: Can this entire test be moved into the one that runs the example/test?

File: tests/docker_run_tests.rs [hunk 3]
```diff
@@ -222,6 +223,53 @@ fn run_postgres() -> Container {
...
+
+    let hash = Uuid::new_v4().to_string();
+    let container_name = format!("mysql-{}", hash);
+    let password = hash;
```
Comment (Milo123459) [inline tests/docker_run_tests.rs:231]: Is there a reason for just making another variable with the same value

**Review by Milo123459 — CHANGES_REQUESTED** [review]

**Comment by FinnDore** [comment]
image
:)

**Comment by FinnDore** [comment]
Nevermind it wont build anymore despite not touching anything really.....

**Comment by FinnDore** [comment]
For postgres it looks like `pg` is required

```
 = note: /nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: cannot find -lpq: No such file or directory
```

**Comment by coffee-cup** [comment]
> For postgres it looks like `pg` is required
> 
> ```
>  = note: /nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: cannot find -lpq: No such file or directory
> ```

We've had to handle postgres specially for both Python and Ruby (e.g. https://github.com/railwayapp/nixpacks/blob/main/src/providers/ruby.rs#L31). Most likely have to do the same here.

**Comment by FinnDore** [comment]
This is not a drill! Mysql is now working.
image

`setup_phase.add_libraries(vec!["mariadb-connector-c".to_owned()]);`

Could someone enlighten me, whats the difference between a package and a library? ( `add_pkgs` vs `add_libraries` ). just the install locations or something?

**Comment by FinnDore** [comment]
Looks like it can find libmysqlclient when i install `libmariadb-dev-compat`, but then it spit this long error out when it trys to build. I have absolutely no idea what this error means either.

```
#16 0.375     Updating crates.io index
#16 24.84  Downloading crates ...
#16 25.47   Downloaded url v1.7.2
#16 25.49   Downloaded diesel v1.4.8
#16 25.55   Downloaded diesel_derives v1.4.1
#16 25.55   Downloaded unicode-ident v1.0.2
#16 25.55   Downloaded quote v1.0.20
#16 25.55   Downloaded proc-macro2 v1.0.40
#16 25.56   Downloaded byteorder v1.4.3
#16 25.57   Downloaded syn v1.0.98
#16 25.59   Downloaded unicode-normalization v0.1.21
#16 25.59   Downloaded tinyvec_macros v0.1.0
#16 25.59   Downloaded pkg-config v0.3.25
#16 25.59   Downloaded matches v0.1.9
#16 25.59   Downloaded percent-encoding v1.0.1
#16 25.59   Downloaded mysqlclient-sys v0.2.5
#16 25.59   Downloaded tinyvec v1.6.0
#16 25.60   Downloaded unicode-bidi v0.3.8
#16 25.60   Downloaded idna v0.1.5
#16 25.63    Compiling proc-macro2 v1.0.40
#16 25.63    Compiling quote v1.0.20
#16 25.63    Compiling unicode-ident v1.0.2
#16 25.63    Compiling tinyvec_macros v0.1.0
#16 25.71    Compiling syn v1.0.98
#16 25.79    Compiling pkg-config v0.3.25
#16 26.28    Compiling matches v0.1.9
#16 26.28    Compiling unicode-bidi v0.3.8
#16 26.29    Compiling percent-encoding v1.0.1
#16 26.35    Compiling byteorder v1.4.3
#16 26.46    Compiling tinyvec v1.6.0
#16 26.60    Compiling mysqlclient-sys v0.2.5
#16 27.22    Compiling unicode-normalization v0.1.21
#16 27.64    Compiling idna v0.1.5
#16 27.95    Compiling url v1.7.2
#16 30.48    Compiling diesel_derives v1.4.1
#16 32.47    Compiling diesel v1.4.8
#16 38.40    Compiling rust-diesel-libmysql v0.1.0 (/app)
#16 38.44 warning: unused imports: `env`, `process::Command`
#16 38.44  --> src/main.rs:1:11
#16 38.44   |
#16 38.44 1 | use std::{env, process::Command};
#16 38.44   |           ^^^  ^^^^^^^^^^^^^^^^
#16 38.44   |
#16 38.44   = note: `#[warn(unused_imports)]` on by default
#16 38.44
#16 39.37 error: linking with `cc` failed: exit status: 1
#16 39.37   |
#16 39.37   = note: "cc" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crt1.o" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crti.o" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtbegin.o" "/tmp/rustcFyXwTe/symbols.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.0.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.1.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.2.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.3.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.4.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.5.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.6.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.rust_diesel_libmysql.a92dd4e1-cgu.7.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745.4sm2f6rnqxbaajh2.rcgu.o" "-Wl,--as-needed" "-L" "/app/target/aarch64-unknown-linux-musl/release/deps" "-L" "/app/target/release/deps" "-L" "/usr/lib/aarch64-linux-gnu" "-L" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib" "-Wl,-Bstatic" "/app/target/aarch64-unknown-linux-musl/release/deps/libdiesel-3f22bd92c4740d12.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/liburl-64c46411e1051d9d.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libpercent_encoding-854f73f5fee739b8.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libidna-679db69d6cd04f70.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libunicode_normalization-01b59102f6f591e4.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libtinyvec-51efa96abdb7667b.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libtinyvec_macros-d632ffde8264a145.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libunicode_bidi-e9776af0cde2a441.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libmatches-6b3d4d911df0191b.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libmysqlclient_sys-af2fe52a143397ee.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libbyteorder-16c82fda548b1fa3.rlib" "-Wl,--start-group" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-02d93f39ea80c989.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libpanic_unwind-689d0a9c2eca58b6.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libobject-220d95266bc37c0b.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libmemchr-8d9327a7c93a2c81.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libaddr2line-b28e615e26724062.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libgimli-c1ef154125dd6069.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_demangle-cbe515a1f227bceb.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd_detect-188796f0cc453cc1.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libhashbrown-86c1895aadf36891.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libminiz_oxide-e0d2bc376054812a.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libadler-10d3b397101b2381.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_alloc-6b783b728bb675ac.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libunwind-c3acab90458e760a.rlib" "-lunwind" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcfg_if-be685a75057d60ea.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liblibc-29c50c82233c0db7.rlib" "-lc" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liballoc-e3c71ab46c563381.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_core-160e4cee98850e94.rlib" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcore-ba779adaa81f1cd7.rlib" "-Wl,--end-group" "/nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-636f1f774ea0846b.rlib" "-Wl,-Bdynamic" "-lmysqlclient" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-nostartfiles" "-L" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib" "-L" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained" "-o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel_libmysql-2ff6b0cda372f745" "-Wl,--gc-sections" "-static" "-no-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtend.o" "/nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtn.o"
#16 39.37   = note: /nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: /nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-02d93f39ea80c989.rlib(std-02d93f39ea80c989.std.143cfb64-cgu.0.rcgu.o): in function `std::sys::unix::os::home_dir::fallback':
#16 39.37           /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/sys/unix/os.rs:625: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#16 39.37           /nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: /nix/store/n94h3m34w0hxy0dwcc2c73cdc54avn9j-rust-std-1.62.1-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-02d93f39ea80c989.rlib(std-02d93f39ea80c989.std.143cfb64-cgu.0.rcgu.o): in function `<std::sys_common::net::LookupHost as core::convert::TryFrom<(&str,u16)>>::try_from':
#16 39.37           /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/sys_common/net.rs:205: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#16 39.37           /nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: /nix/store/rkx7gbjzjmxl1bizgfc6hh4c98vjdrg4-rust-default-1.62.1/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/libunwind.a(libunwind.o): undefined reference to symbol 'dladdr@@GLIBC_2.34'
#16 39.37           /nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: /nix/store/jq8i4896sgmgbkk7nbjrnidxpxmgcid5-glibc-2.34-210/lib/libc.so.6: error adding symbols: DSO missing from command line
#16 39.37           collect2: error: ld returned 1 exit status
#16 39.37
#16 39.37   = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
#16 39.37   = note: use the `-l` flag to specify native libraries to link
#16 39.37   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
#16 39.37
#16 39.38 warning: `rust-diesel-libmysql` (bin "rust-diesel-libmysql") generated 1 warning
#16 39.38 error: could not compile `rust-diesel-libmysql` due to previous error; 1 warning emitted
------
executor failed running [/bin/bash -ol pipefail -c cargo build --release --target aarch64-unknown-linux-musl]: exit code: 101
test test_rust_diesel ... FAILED
```

**Comment by FinnDore** [comment]
For what its worth installing `libmariadb-dev` using apt-get resolves this issue when using the container `debian:buster-slim`. https://github.com/FinnDore/something/blob/a4a1f558d189dea3700c046fa4d48899f9ccb43f/apps/something/dockerfile#L12

**Comment by FinnDore** [comment]
Looks like some of the rust tests broke in that merge. shall investigate tomorrow :)

**Comment by coffee-cup** [comment]
> For what its worth installing libmariadb-dev using apt-get resolves this issue when using the container debian:buster-slim.

If the package is not available on Nix then feel free to install from apt

**Comment by FinnDore** [comment]
The current changes attempt to install `libmariadb-dev-compat` from apt, I have also just tried to use `libmariadb-dev` from apt but still no luck. ( providing I'm actually installing them correctly. ). I will have a try this week to build the image without the failing steps and see if I can manually get it working. It must be possible!

**Comment by coffee-cup** [comment]
Apologizes that I haven't had a chance to take over this PR yet. I'm going to close this for now since it has gone stale, but will come back to this issue at a later date.

**Comment by FinnDore** [comment]
Yea no worries, pretty much the same from my side. sounds good to me.

**Pull request closed without merging**